File: src/main/java/com/gw2auth/oauth2/server/service/client/registration/ClientRegistrationService.java
package com.gw2auth.oauth2.server.service.client.registration;

import java.util.*;

public interface ClientRegistrationService {

    List<ClientRegistration> getClientRegistrations(UUID accountId);

    Optional<ClientRegistration> getClientRegistration(UUID accountId, UUID id);

    List<ClientRegistration> getClientRegistrations(Collection<UUID> ids);

    Optional<ClientRegistration> getClientRegistration(UUID id);

    ClientRegistrationCreation createClientRegistration(UUID accountId, String displayName, Set<String> authorizationGrantTypes, Set<String> redirectUris);

    ClientRegistration addRedirectUri(UUID accountId, UUID id, String redirectUri);

    ClientRegistration removeRedirectUri(UUID accountId, UUID id, String redirectUri);

    ClientRegistrationCreation regenerateClientSecret(UUID accountId, UUID id);

    void deleteClientRegistration(UUID accountId, UUID id);
}

File: src/main/angular/app/main/account/client/client-registration.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {
  ClientRegistrationCreation,
  ClientRegistrationCreationRequest,
  ClientRegistrationPrivate,
  ClientRegistrationPrivateSummary
} from './client-registration.model';

@Injectable()
export class ClientRegistrationService {

  constructor(private readonly http: HttpClient) { }

  getClientRegistrations(): Observable<ClientRegistrationPrivate[]> {
    return this.http.get<ClientRegistrationPrivate[]>('/api/client/registration');
  }

  getClientRegistration(clientId: string): Observable<ClientRegistrationPrivate> {
    return this.http.get<ClientRegistrationPrivate>('/api/client/registration/' + encodeURIComponent(clientId));
  }

  getClientRegistrationSummary(clientId: string): Observable<ClientRegistrationPrivateSummary> {
    return this.http.get<ClientRegistrationPrivateSummary>('/api/client/registration/' + encodeURIComponent(clientId) + '/summary');
  }

  createClientRegistration(request: ClientRegistrationCreationRequest): Observable<ClientRegistrationCreation> {
    return this.http.post<ClientRegistrationCreation>('/api/client/registration', request);
  }

  addRedirectUri(clientId: string, redirectUri: string): Observable<ClientRegistrationPrivate> {
    return this.http.put<ClientRegistrationPrivate>('/api/client/registration/' + encodeURIComponent(clientId) + '/redirect-uris', redirectUri);
  }

  removeRedirectUri(clientId: string, redirectUri: string): Observable<ClientRegistrationPrivate> {
    return this.http.delete<ClientRegistrationPrivate>('/api/client/registration/' + encodeURIComponent(clientId) + '/redirect-uris', { params: { redirectUri: redirectUri } });
  }

  regenerateClientSecret(clientId: string): Observable<ClientRegistrationCreation> {
    return this.http.patch<ClientRegistrationCreation>('/api/client/registration/' + encodeURIComponent(clientId) + '/client-secret', null);
  }

  deleteClientRegistration(clientId: string): Observable<void> {
    return this.http.delete<void>('/api/client/registration/' + encodeURIComponent(clientId));
  }
}
